src/FilesystemSyncTaskHelper.php

```diff
@@ -0,0 +1,349 @@
+                $file = $this->flysystemAssetStore::getFileByFilename($fileTuple['path']);
+                if ($file) {
+                    $this->results['filesSkippedFromFilesystem']++;
+                    return false;
+                }
+
+                return true;
+            }
+
+            $file = File::find($fileTuple['path']);
+            if ($file) {
+                $this->results['filesSkippedFromFilesystem']++;
+                return false;
+            } elseif ($this->delete_broken) {
+                // file exists in filesystem but not DB and is in a hash folder
+                $contents = file_get_contents($realPath);
+                $sha1 = sha1($contents);
+
+                $parts = explode(DIRECTORY_SEPARATOR, $fileTuple['path']);
+                $parent = array_pop($parts);
+                if (substr($sha1, 0, 10) == $parent) {
+                    unlink($realPath);
+                    $this->dirsToRemove[]=dirname($realPath);
+                    $this->results['filesRemovedFromFilesystem']++;
+                    return false;
+                }
+            }
+
+            // require extensions for files
+            $extension = pathinfo($fileTuple['path'], PATHINFO_EXTENSION);
+            if (!$extension) {
+                return false;
+            }
+
+            // require file to use allowed extension
+            if (!in_array($extension, $this->allowedExtensions)) {
+                return false;
+            }
+
+            // check if parent is a hash
+            $path = dirname($fileTuple['path']);
+            $exploded = explode(DIRECTORY_SEPARATOR, $path);
+            $parent = array_pop($exploded);
+            $table = DataObject::singleton(File::class)->baseTable();
+            $parentExists = DB::prepared_query("SELECT COUNT(*) FROM \"$table\" WHERE \"FileHash\" LIKE ?", ["$parent%"])->value();
+            if ($parentExists) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Sync the provided file with the database.
+     * @param array $fileTuple
+     * @throws \SilverStripe\ORM\ValidationException
+     */
+    private function syncFileToDatabase(array $fileTuple)
+    {
+        $realPath = $this->publicAssetAdapter->relativeToRealPath($fileTuple['path']);
+        $relativePath = substr($realPath, strlen($this->publicAssetAdapter->getPathPrefix()));
+        $relativeParentPath = dirname($relativePath);
+
+        // Create the file
+        if ($fileTuple['type'] == 'dir') {
+            /** @var Folder $file */
+            $file = Injector::inst()->create(Folder::class);
+            $file->setFilename(basename($fileTuple['path']));
+        } else {
+            $fileClass = File::get_class_for_file_extension(pathinfo($fileTuple['path'], PATHINFO_EXTENSION));
+            /** @var File $file */
+            $file = Injector::inst()->create($fileClass);
```

Comment:
`Folder` is also a subclass of `File`.

I would have initialised my `$file` variable before going into the `if` statement.

```php
/** @var Folder $file */
$file = null;
```